Adding the workflow creation logic as well as a thin server.

cchristelis opened the pull request.

martinslabber commented on src/katsdpk8spoc/controller_server.py:
We should read in a config file.
It will be important to get the namespace per subarray from config. Since multiple people will use the same cluster for testing I'd like it to get namespaces from config.

[default]
argo_url=blabla
argo_token_file=blabla
subarrays=4
serviceaccount=workflow

[subarray1]
k8s_namespace=sdparray1

...

martinslabber replied:
OR maybe the config should be yaml then it will display nice if we make it a configmap object in Kubernetes. 

---
argo_url: blabla
argo_token_file: blabla
serviceaccount: workflow
subarrays:
    - subarray1:
         k8s_namespace: sdparray1
    - subarray2:
         k8s_namespace: sdparray2

cchristelis replied:
This will be followed up in https://github.com/ska-sa/katsdpk8spoc/issues/4

cchristelis replied:
This will be followed up in https://github.com/ska-sa/katsdpk8spoc/issues/4

martinslabber commented on src/katsdpk8spoc/workflow_controller.py:
```diff
@@ -0,0 +1,203 @@
+        self.when = f"{{tasks.batch-setup.outputs.result}} >= {number}"
+
+
+class ProductControllerWorkflow:
```
This is good for now.
Passing in config and picking namespace and service account from config will be based on subarray will be needed.

cchristelis replied:
This will be followed up in https://github.com/ska-sa/katsdpk8spoc/issues/5

martinslabber commented on src/katsdpk8spoc/workflow_controller.py:
```diff
@@ -0,0 +1,203 @@
+        """
+        ingest_count = worker_count // 5
+        calib_count = worker_count // 4
+        self.tasks = [Telstate()]
+        self.tasks += [Ingest(n + 1) for n in range(ingest_count)]
+        self.tasks += [Calibrator(n + 1) for n in range(calib_count)]
+        self.tasks += [BatchSetup()]
+        self.tasks += [Batch(n + 1) for n in range(worker_count)]
```
I like this

martinslabber reviewed (changes requested):
I have made lots of comments that can be addressed in follow up PRs.

But now we need.

1. The index file and using the safe loading of it. It is broken without that.
2. A config file that allows us to set the Argo base URL and Argo token file. We cannot start with the existence of argotoken.py.

Thank you for the Workflow generator, it looks really good.

martinslabber commented on src/katsdpk8spoc/argotoken.py:
```diff
@@ -0,0 +1,2 @@
+ARGO_BASE_URL = "http://10.8.103.151:2746"
```
See below, I think just passing in a single config filename on the command line will be better.

cchristelis commented on src/katsdpk8spoc/controller_server.py:
done

cchristelis commented on src/katsdpk8spoc/controller_server.py:
This will be followed up in https://github.com/ska-sa/katsdpk8spoc/issues/3

cchristelis commented on src/katsdpk8spoc/controller_server.py:
Will be followed up in https://github.com/ska-sa/katsdpk8spoc/issues/3

cchristelis commented:
I've addressed most of your requests @martinslabber. There were some that were a little too big to deal with as part of this PR. These big changes I've added as new tickets.
